spec/controllers/course/statistics/aggregate_controller_spec.rb

```diff
@@ -210,5 +210,77 @@
         it { expect(subject).to be_successful }
       end
     end
+
+    describe '#submission_due' do
+      render_views
+      let!(:deadline) { 1.day.from_now }
+      let!(:published_assessment) do
+        create(:assessment, :published, :with_all_question_types, course: course, end_at: deadline)
+      end
+
+      let!(:unpublished_assessment) { create(:assessment, :with_all_question_types, course: course) }
+      let!(:student) { create(:course_student, course: course) }
+
+      let!(:non_submitted_submission) do
+        create(:submission, :attempting,
+               assessment: published_assessment, creator: student.user,
+               created_at: 1.hour.from_now)
+      end
+
+      subject do
+        get :submission_due, format: :json,
+                             params: { course_id: course, user_id: course_user, student_id: student.id }
+      end
+
+      context 'when a Normal User pings the endpoint' do
+        let(:user) { create(:user) }
+        before { sign_in(user) }
+        it { expect { subject }.to raise_exception(CanCan::AccessDenied) }
+      end
+
+      context 'when a Course Student pings the endpoint' do
+        let(:user) { create(:course_student, course: course).user }
+        before { sign_in(user) }
+        it { expect { subject }.to raise_exception(CanCan::AccessDenied) }
+      end
+
+      context 'when a Course Teaching Assistant pings the endpoint' do
+        let(:user) { create(:course_teaching_assistant, course: course).user }
+        before { sign_in(user) }
+        it { expect(subject).to be_successful }
+      end
+
+      context 'when a Course Manager pings the endpoint' do
+        let(:user) { create(:course_manager, course: course).user }
+        before { sign_in(user) }
+
+        it 'expects to render all published assessments' do
+          expect(subject).to be_successful
+          json_result = JSON.parse(response.body)
+          expect(json_result['assessments'].count).to eq(1)
+
+          expected_end_at = published_assessment.end_at.change(usec: published_assessment.end_at.usec / 1000 * 1000)
+          parsed_end_at = DateTime.parse(json_result['assessments'][0]['endAt']).
+                          change(usec: DateTime.parse(json_result['assessments'][0]['endAt']).
+                                       usec / 1000 * 1000)
+          parsed_reference_end_at = DateTime.parse(json_result['assessments'][0]['referenceEndAt']).
+                                    change(usec: DateTime.parse(json_result['assessments'][0]['referenceEndAt']).
+                                                 usec / 1000 * 1000)
+
+          expect(json_result['assessments'][0]['workflowState']).to eq('attempting')
+          expect(parsed_end_at).to eq(expected_end_at)
+          expect(parsed_reference_end_at).to eq(expected_end_at)
```

Comment:
isn't there a `.to be_within(time_delta)` method that we can use instead of doing truncation like this?

Reply:
you're right. I have done some more inspection and indeed, previously I was using this `be_within` function but I was doing it wrongly and thus it's unsuccessful

now that I have done it properly, it gets right now